# Fire Image Classification

## Imports

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
# Keras Imports
import tensorflow as tf
import keras
from keras import backend as K
# CNN and MLP architecture
from keras.models import Sequential
from keras.layers import (
    Dense,
    Conv2D,
    MaxPooling2D,
    UpSampling2D,
    Dropout,
    Flatten,
    BatchNormalization
)
from keras.models import Model
from keras.optimizers import SGD
from keras.initializers import RandomNormal
# Keras Callbacks
from keras.callbacks import EarlyStopping, TensorBoard
# Time Training
import time
# Data Splitting
from sklearn.model_selection import train_test_split
# Image Preprocessing
from PIL import Image
from keras.preprocessing.image import img_to_array

Using TensorFlow backend.


## Get the Data

In [2]:
df = pd.read_csv('Fire-Detection-Image-Dataset/fires.csv')

In [3]:
df.head()

Unnamed: 0           Folder  \
0           0  Normal Images 3   
1           1  Normal Images 3   
2           2  Normal Images 3   
3           3  Normal Images 3   
4           4  Normal Images 3   

                                            filename  label  
0  Hotel_Monterey_La_Soeur_Osaka_standard_twin_be...      0  
1                                         house5.jpg      0  
2                                 mi-plage-hawai.jpg      0  
3  JB224_03_Colourful_christmas_table_setting_in_...      0  
4                      Interior Design Ideas (3).jpg      0

In [4]:
# define predictor and target sets of data
features = [df.columns[1], df.columns[2]]
target = [df.columns[3]]
X, y = df[features], df[target]
# Store number of class
num_classes = 2

## Split DataFrame Based on Label

In [5]:
# Split df into train and test, based on label
# Credit goes to this Stack Overflow answer for the following:
# https://stackoverflow.com/questions/24147278/how-do-i-create-test-and-train-samples-from-one-dataframe-with-pandas
selector = np.random.rand(len(df)) < 0.75
df_train, df_test = df[selector], df[~selector]
print(f"Number of Training Samples: {len(df_train)}")
print(f"Number of Testing Samples: {len(df_test)}")

Number of Training Samples: 508
Number of Testing Samples: 143


## Defining the Optimal Model 

### Build a Generator

For this dataset, we'll increase the efficiency of training by loading in subsections of the dataset to train on at a time, using Python-esque generator in Keras!

In [6]:
def data_gen(df_gen, batch_size):
    """Generate batches of the dataset to train the model on, one subsection at a time.
       Credit goes to Milad Toutounchian for this implementation, originally found at:
       https://github.com/Make-School-Courses/DS-2.2-Deep-Learning/blob/master/Final_Project/image_data_prep.ipynb
    
       Parameters:
       df(DataFrame): larger portion of the datset used for training
       batch_size(int): the number of samples to include in each batch
       
       Returns:
       tuple: input features of the batch, along with corresponding labels
    
    """
    while True:
        # list of images
        x_batch = np.zeros((batch_size, 1024, 1024, 3))
        # list of labels
        y_batch = np.zeros((batch_size, 1))
        # add samples until we reach batch size
        for j in range(len(df_gen) // batch_size):
            batch_index = 0
            for index in df_gen['Unnamed: 0']:
                # add image to the input
                filepath = f"Fire-Detection-Image-Dataset/{df_gen['Folder'][index]}/{df_gen['filename'][index]})
                img = Image.open(filepath)
                image_red = img.resize((1024, 1024))
                x_batch[batch_index] = img_to_array(image_red)
                # set label
                y_batch[batch_index] = df_gen['label'][index]
                # increment index in the batch
                batch_index += 1
            yield (x_batch, y_batch)

### Finding the Best Architecture

Now we'll find the optimal number of layers and neurons per layer using TensorBoard.

In [7]:
# Helper functions to Reduce Repetition of Adding Layers
def add_conv_layer(model, layer_size, needs_input):
    """Add a Keras convolutional layer to the model, along with MaxPooling.
       Will specify input shape as well if needed.
       
       Parameters:
       model(Model): Neural network in Keras
       layer_size(int): number of neurons to go in layer
       need_input(bool): signals if the convolutional layer needs to specify
                         the dimensions of the input
       
       Returns: None
       
    """
    if needs_input is True:
        # specify input dimension for 1st conv layer
        conv_layer = Conv2D(layer_size,
                            kernel_size=(3, 3),
                            activation='relu',
                            input_shape=(1024, 1024, 3))

    else:
        # otherwise all other convolutional layers don't need it
        conv_layer = Conv2D(layer_size,
                            kernel_size=(3, 3),
                            activation='relu')
    # add Convolutional layer
    model.add(conv_layer)  
    # add MaxPooling layer
    model.add(MaxPooling2D(pool_size=(2, 2)))  # no learning params
    return None
                  
    
def add_dense_layer(model, layer_size, is_output, drop_rate):
    """Add a multi-layer perceptron to the model
       Will specify 'sigmoid' for the final layer.
       
       Parameters:
       model(Model): Neural network in Keras
       layer_size(int): number of neurons to go in layer
       is_output(bdool): signals if the MLP is the last layer
       drop_rate(float): percentage of connections in Dense layer
                       to cut off
       
       Returns: None
       
    """
    # specify activation function
    activation = 'relu' if is_output is False else 'sigmoid'
    # add MLP
    model.add(Dense(layer_size, activation=activation)) 
    # Add Dropout layer and Batch Normalization
    if is_output is False:
        model.add(BatchNormalization())
        model.add(Dropout(drop_rate))
    return None


def define_model(units, conv_layers, dense_layers, dropout):
    """Define a CNN + MLP model in Keras.
    
       Parameters:
       units(int): number of neurons to go in a layer
       conv_layers(int): number of convolutional layers
       dense_layers(int): number of MLP
       dropout(float): percentage of connections in Dense layer
                       to cut off
                       
       Returns: tf.keras.Sequential: the neural network to train
    
    """
    # Instaniate model
    model = Sequential()
    # Add CNN layers
    add_conv_layer(model, units, True)
    for l in range(conv_layers - 1):
        # add convolutional layers that come after the 1st
        add_conv_layer(model, units, False)
    # Flatten the data
    model.add(Flatten())
    # Add MLP Layers
    for l in range(dense_layers - 1):
        add_dense_layer(model, units, False, dropout)
    # add final MLP, for output
    add_dense_layer(model, 1, True, dropout)
    # Compile Model
    model.compile(loss=keras.losses.binary_crossentropy,
                  optimizer=keras.optimizers.Adadelta(),
                  metrics=['accuracy',
                           tf.keras.metrics.Precision(),
                           tf.keras.metrics.Recall()])
    # Return model
    return model

In [8]:
# Choices for the Model Architecture - values arbitrary
dense_layers = [1, 2, 3]
layer_sizes = [4, 8, 16]
conv_layers = [1, 2, 3]

# try different combinations!
# These for loops come from https://youtu.be/lV09_8432VA
for dense_layer in dense_layers:
    for size in layer_sizes:
        for conv in conv_layers:
            # name the combo
            NAME = (
                f'{conv}-conv-{size}-nodes' +
                f'-{dense_layer}-dense_layers' + 
                f'-{int(time.time())}'
            )
            # Instantiate TensorBoard to visualize model performance
            tensorboard = TensorBoard(log_dir=f'./Graph/{NAME}')
            # Define Model
            model = define_model(size, conv, dense_layer, 0.2)
            # Train the Model (using a generator!)
            epochs, batch_size = 5, 20
            history = model.fit_generator(generator=data_gen(df_train, batch_size=batch_size),
                                steps_per_epoch=len(df_train['label']) // batch_size,
                                epochs=epochs,
                                validation_data=data_gen(df_test, batch_size=batch_size),
                                validation_steps=len(df_test['label']) // batch_size, 
                                callbacks=[tensorboard])

Instructions for updating:
If using Keras pass *_constraint arguments to layers.

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Epoch 1/5
25/25 [==============================] - 58s 2s/step - loss: 1545.5610 - accuracy: 0.9140 - precision: 0.1460 - recall: 0.1331 - val_loss: 0.0000e+00 - val_accuracy: 1.0000 - val_precision: 0.9051 - val_recall: 0.8731

Epoch 2/5
25/25 [==============================] - 60s 2s/step - loss: 1266.1505 - accuracy: 0.9100 - precision: 0.8779 - recall: 0.8885 - val_loss: 0.0000e+00 - val_accuracy: 1.0000 - val_precision: 0.9315 - val_recall: 0.9118
Epoch 3/5
25/25 [==============================] - 65s 3s/step - loss: 1196.4379 - accuracy: 0.9220 - precision: 0.9172 - recall: 0.9149 - val_loss: 0.0000e+00 - val_accuracy: 1.0000 - val_precision: 0.9406 - val_recall: 0.9301
Epoch 4/5
25/25 [==============================] - 60s 2s/step - loss: 932.5430 - accuracy: 0.9220 - precision: 0.9306 - recall: 0.9297 

25/25 [==============================] - 111s 4s/step - loss: 9.2137 - accuracy: 0.9220 - precision_5: 0.9157 - recall_5: 0.8987 - val_loss: 5.6039e-10 - val_accuracy: 1.0000 - val_precision_5: 0.9485 - val_recall_5: 0.9254
Epoch 3/5
25/25 [==============================] - 109s 4s/step - loss: 5.7428 - accuracy: 0.9200 - precision_5: 0.9341 - recall_5: 0.9251 - val_loss: 3.8749e-06 - val_accuracy: 1.0000 - val_precision_5: 0.9499 - val_recall_5: 0.9352
Epoch 4/5
25/25 [==============================] - 113s 5s/step - loss: 4.4470 - accuracy: 0.9200 - precision_5: 0.9402 - recall_5: 0.9339 - val_loss: 1.1152e-04 - val_accuracy: 1.0000 - val_precision_5: 0.9506 - val_recall_5: 0.9397
Epoch 5/5
25/25 [==============================] - 118s 5s/step - loss: 3.5241 - accuracy: 0.9200 - precision_5: 0.9432 - recall_5: 0.9384 - val_loss: 0.0032 - val_accuracy: 1.0000 - val_precision_5: 0.9509 - val_recall_5: 0.9424
Epoch 1/5
25/25 [==============================] - 113s 5s/step - loss: 4200.7

25/25 [==============================] - 89s 4s/step - loss: 0.7357 - accuracy: 0.4880 - precision_12: 0.3011 - recall_12: 0.2989 - val_loss: 0.0160 - val_accuracy: 1.0000 - val_precision_12: 0.5061 - val_recall_12: 0.4527
Epoch 3/5
25/25 [==============================] - 127s 5s/step - loss: 0.7200 - accuracy: 0.4920 - precision_12: 0.5090 - recall_12: 0.4984 - val_loss: 0.3418 - val_accuracy: 1.0000 - val_precision_12: 0.5871 - val_recall_12: 0.5360
Epoch 4/5
25/25 [==============================] - 101s 4s/step - loss: 0.7213 - accuracy: 0.4740 - precision_12: 0.5758 - recall_12: 0.5542 - val_loss: 0.6834 - val_accuracy: 0.6000 - val_precision_12: 0.6109 - val_recall_12: 0.5446
Epoch 5/5
25/25 [==============================] - 107s 4s/step - loss: 0.6968 - accuracy: 0.5540 - precision_12: 0.6013 - recall_12: 0.5443 - val_loss: 0.4042 - val_accuracy: 1.0000 - val_precision_12: 0.6386 - val_recall_12: 0.5615
Epoch 1/5
25/25 [==============================] - 117s 5s/step - loss: 0.7

Epoch 2/5
25/25 [==============================] - 79s 3s/step - loss: 0.6945 - accuracy: 0.4420 - precision_19: 0.5534 - recall_19: 0.3292 - val_loss: 0.6247 - val_accuracy: 1.0000 - val_precision_19: 0.6556 - val_recall_19: 0.3038
Epoch 3/5
25/25 [==============================] - 86s 3s/step - loss: 0.6937 - accuracy: 0.5220 - precision_19: 0.6813 - recall_19: 0.3330 - val_loss: 0.6176 - val_accuracy: 1.0000 - val_precision_19: 0.7582 - val_recall_19: 0.3683
Epoch 4/5
25/25 [==============================] - 85s 3s/step - loss: 0.6927 - accuracy: 0.5240 - precision_19: 0.7045 - recall_19: 0.3884 - val_loss: 0.6100 - val_accuracy: 1.0000 - val_precision_19: 0.7613 - val_recall_19: 0.4370
Epoch 5/5
25/25 [==============================] - 91s 4s/step - loss: 0.6920 - accuracy: 0.5220 - precision_19: 0.6982 - recall_19: 0.4566 - val_loss: 0.6032 - val_accuracy: 1.0000 - val_precision_19: 0.7391 - val_recall_19: 0.5064
Epoch 1/5
25/25 [==============================] - 99s 4s/step - los

Epoch 2/5
25/25 [==============================] - 126s 5s/step - loss: 0.7502 - accuracy: 0.5220 - precision_26: 0.4447 - recall_26: 0.3461 - val_loss: 2.4685 - val_accuracy: 0.0000e+00 - val_precision_26: 0.5508 - val_recall_26: 0.3776
Epoch 3/5
25/25 [==============================] - 140s 6s/step - loss: 0.7368 - accuracy: 0.5260 - precision_26: 0.5042 - recall_26: 0.3630 - val_loss: 1.2468 - val_accuracy: 0.1500 - val_precision_26: 0.5670 - val_recall_26: 0.3820
Epoch 4/5
25/25 [==============================] - 132s 5s/step - loss: 0.7425 - accuracy: 0.5280 - precision_26: 0.5380 - recall_26: 0.3808 - val_loss: 0.6805 - val_accuracy: 0.4500 - val_precision_26: 0.5874 - val_recall_26: 0.4072
Epoch 5/5
25/25 [==============================] - 124s 5s/step - loss: 0.7175 - accuracy: 0.5020 - precision_26: 0.5648 - recall_26: 0.4142 - val_loss: 1.1141 - val_accuracy: 0.1500 - val_precision_26: 0.5919 - val_recall_26: 0.4193


**Final Conclusion: What's the Best Architecture for the Model?**

After training the different combinations above and looking through the [logs](Graph/) returned by Tensorboard, the best model archtecture has been determined to be:

- 3 Conv2D Layers
- 1 Dense Layer
- 4 Neurons in each Layer
- Dropout Rate of 20%
- 5 epochs
- 20 samples per Batch

This model had the following metrics, which was the best of any of the other models:

- Validation Accuracy: 100%
- Validation Loss: 4.51e31%
- Validation Precision: 96.25%
- Validation Recall: 95.09%
- F1-Score: 0.9567

Although the accuracy of 100% is a little concerning, it may be possible to prevent overfitting by increase the drop out rate to 30%.

###  Finding Optimal Hyperparameters

Now, we'll find the optimal values for the hyperparameters using hyperas.

### Summarize the Model

In [5]:
print(model.summary())

Model: "sequential_9"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_17 (Conv2D)           (None, 26, 26, 32)        320       
_________________________________________________________________
conv2d_18 (Conv2D)           (None, 24, 24, 64)        18496     
_________________________________________________________________
max_pooling2d_9 (MaxPooling2 (None, 12, 12, 64)        0         
_________________________________________________________________
flatten_9 (Flatten)          (None, 9216)              0         
_________________________________________________________________
dense_17 (Dense)             (None, 128)               1179776   
_________________________________________________________________
dense_18 (Dense)             (None, 2)                 258       
Total params: 1,198,850
Trainable params: 1,198,850
Non-trainable params: 0
____________________________________________

## Upsample the Minority Class

We need more fire!

## Data Augumentation

Let's add more variation to the dataset now, in order to make the model more robust!

## Model Evaluation

## Final Conclusions